sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -105,6 +105,17 @@ class RocksDB(
   }
 
   columnFamilyOptions.setCompressionType(getCompressionType(conf.compression))
+  // We can easily accumulate many small L0 files if changelog is not enabled, or full snapshot
+  // interval is not large enough. Triggering compactions for those small files are expensive.
+  // We make it harder to trigger and allow more L0 files without write stalling.
+  // We increase L0->L1 compaction threshold to reduce the overhead of L0->L1 compaction.
+  // Given the nature of small L0 files, for most workloads, even the value of 16 is too low for
+  // L0->L1 write amplification. However, if the value is too large, we may risk the chance that
+  // in some workloads where batch size is very large, some data might take a very long time to
+  // be compacted.
+  columnFamilyOptions.setLevel0FileNumCompactionTrigger(16)
+  columnFamilyOptions.setLevel0SlowdownWritesTrigger(200)
```

Comment:
Do we need to make any of these configurable ?

Reply:
It is the crucial part of the PR. By default we do snapshot checkpoint every 10 batches, and compaction triggers in every 4 flushes (right now a snapshot checkpoint does an implicit flush). So in 1 in 40 batches, we will trigger a compaction, and that batch's latency will impact P99. If we increase the compaction threshold from 4 to 16, we will only do compaction every 1 in 160 batches, and it will push this out of P99.

Of course, by having it to happen 1/4 of the time, the compaction is much less likely to catchup, so we probably won't have the problem anyway, but even if we have the problem in another workload, it is much less likely to impact P99.